R/estimate_KM.R

```diff
@@ -86,39 +82,51 @@ estimate_KM <- function(
   if (!is.data.frame(data))
     stop("Data does not have class `data.frame`.")
 
-  reqcols <- c(strata, CNSR, AVAL)
-  if (! all(reqcols %in% colnames(data))){
-    stop(paste0("Following columns are missing from `data`: ", paste(setdiff(reqcols, colnames(data)), collapse = " "), "."))
+  if (is.null(formula)) {
+    reqcols <- c(strata, CNSR, AVAL)
+    if (! all(reqcols %in% colnames(data))){
+      stop(paste0("Following columns are missing from `data`: ", paste(setdiff(reqcols, colnames(data)), collapse = " "), "."))
+    }
+
+    if (!is.numeric(data[[AVAL]])){
+      stop("Analysis variable (AVAL) is not numeric.")
+    }
+
+    if (!is.numeric(data[[CNSR]])){
+      stop("Censor variable (CNSR) is not numeric.")
+    }
+  }
+  else if (!inherits(formula, "formula")) {
+    stop("Argument `formula=` must be class 'formula'.")
   }
 
-  if (!is.numeric(data[[AVAL]])){
-    stop("Analysis variable (AVAL) is not numeric.")
+  # add strata object if user passer formula -----------------------------------
+  if (!is.null(formula)) {
+    # extract strata
+    formula_rhs <- formula
+    rlang::f_lhs(formula_rhs) <- NULL
+    strata <-
+      stats::get_all_vars(formula = formula_rhs, data = data) %>%
+      names() %>%
+      switch(!rlang::is_empty(.), .) # convert empty string to NULL
   }
 
-  if (!is.numeric(data[[CNSR]])){
-    stop("Censor variable (CNSR) is not numeric.")
+  # construct formula if not passed by user ------------------------------------
+  if (is.null(formula)) {
+    formula <-
+      formula <- stats::as.formula(paste0("survival::Surv(", AVAL, ", 1-", CNSR, ") ~ ",
+                                          ifelse(is.null(strata), "1", paste(strata, collapse = " + "))))
   }
 
   # Remove NA from the analysis ------------------------------------------------
-
-  data <-
-    as.data.frame(data) %>%
-    tidyr::drop_na(any_of(c(AVAL, CNSR)))
-
-  if (!is.null(strata)){
-    data <- data %>%
-      tidyr::drop_na(any_of(strata))
-  }
+  data <- tidyr::drop_na(data, dplyr::all_of(all.vars(formula)))
 
   # Ensure the presence of at least one strata -----------------------------
 
   formula_rhs <-
     ifelse(is.null(strata), "1", paste(strata, collapse = " + "))
 
   # Calculate survival and add time = 0 to survfit object -------------------
-
-  formula <- stats::as.formula(paste0("survival::Surv(", AVAL, ", 1-", CNSR, ") ~ ", formula_rhs))
```

Comment:
We were going to keep this to ensure the prefix is kept in the call right?

Reply:
we're exporting the Surv function now (to facilitate use for the formula argument), and the prefix is not needed. it is still added if a user utilized the CDISC args.

Reply:
I would still prefer to have the prefix. It is part of our good programming practices in the package.

Reply:
perhaps I misunderstood your first comment. this exact line of code appears on line 140